components/builder-originsrv/src/migrations/origin_channels.rs

```diff
@@ -284,5 +284,92 @@ pub fn migrate(migrator: &mut Migrator) -> SrvResult<()> {
+                          INNER JOIN origins o on oc.origin_id = o.id
+                          WHERE o.name = op_origin
+                          AND oc.name = op_channel
+                          AND op.ident LIKE (op_ident  || '%')
+                          AND (op.visibility='public' OR (op.visibility IN ('private', 'hidden') AND op.origin_id IN (SELECT origin_id FROM origin_members WHERE account_id = op_account_id)))
+                          ORDER BY ident ASC
+                          LIMIT op_limit OFFSET op_offset;
+                        RETURN;
+                    END
+                    $$ LANGUAGE plpgsql STABLE"#,
+    )?;
+    migrator.migrate(
+        "originsrv",
+        r#"CREATE OR REPLACE FUNCTION get_origin_channel_package_latest_v4 (
+                    op_origin text,
+                    op_channel text,
+                    op_ident text,
+                    op_target text,
+                    op_account_id bigint
+                 ) RETURNS SETOF origin_packages AS $$
+                    BEGIN
+                        RETURN QUERY SELECT op.*
+                          FROM origin_packages op
+                          INNER JOIN origin_channel_packages ocp on ocp.package_id = op.id
+                          INNER JOIN origin_channels oc on ocp.channel_id = oc.id
+                          INNER JOIN origins o on oc.origin_id = o.id
+                          WHERE o.name = op_origin
+                          AND oc.name = op_channel
+                          AND (op.visibility='public' OR (op.visibility IN ('private', 'hidden') AND op.origin_id IN (SELECT origin_id FROM origin_members WHERE account_id = op_account_id)))
+                          AND op.ident LIKE (op_ident  || '%')
+                          AND op.target = op_target;
+                        RETURN;
+                    END
+                    $$ LANGUAGE plpgsql STABLE"#,
+    )?;
+    migrator.migrate(
+        "originsrv",
+        r#"CREATE OR REPLACE FUNCTION get_origin_channel_package_v3 (
+                    op_origin text,
+                    op_channel text,
+                    op_ident text,
+                    op_account_id bigint
+                 ) RETURNS SETOF origin_packages AS $$
+                    BEGIN
+                        RETURN QUERY SELECT op.*
+                          FROM origin_packages op
+                          INNER JOIN origin_channel_packages ocp on ocp.package_id = op.id
+                          INNER JOIN origin_channels oc on ocp.channel_id = oc.id
+                          INNER JOIN origins o on oc.origin_id = o.id
+                          WHERE op.ident = op_ident
+                          AND o.name = op_origin
+                          AND oc.name = op_channel
+                          AND (op.visibility='public' OR (op.visibility IN ('private', 'hidden') AND op.origin_id IN (SELECT origin_id FROM origin_members WHERE account_id = op_account_id)));
```

Comment:
Wondering if it might be easier to express this (and better performing) using a `LEFT JOIN` on the member table, and then selecting where the package visibility is public OR the membership account_id is not null.

Actually, looking at the rest of the functions in this PR, it might be worth trying to encapsulate the logic for visibility into a view, so it can be reused.

Reply:
I think that's a great idea, but given our current time constraints, I think it might be better to make an issue for that and come back to it at a later time.

Reply:
Yup, sounds good.